ibis/expr/operations/core.py

```diff
@@ -114,14 +117,80 @@ def flat_args(self):
 
 @public
 class ValueOp(Node):
+    # columnar if any of the arguments is column-like
+    output_shape = rlz.shape_like("args")
+
+    # TODO(kszucs): we may want to enable this by default instead of
+    # requiring to implement output_dtype abstract property
+    # data type which all of the arguments are implicitly castable to
+    # output_dtype = rlz.dtype_like("args")
+
     def root_tables(self):
         return distinct_roots(*self.exprs)
 
-    def resolve_name(self):
-        raise ExpressionError(f'Expression is not named: {type(self)}')
+    # def resolve_name(self):
+    #     raise ExpressionError(f'Expression is not named: {type(self)}')
+
+    @property
+    @abstractmethod
+    def output_dtype(self):
+        ...
+
+    @property
+    def output_type(self):
```

Comment:
Add comments explaining that we defer the type selection to the `output_dtype` and `output_shape` properties.

Reply:
Bump

Reply:
Added, but not extensively.